### PR TITLE
greg/NFSE-5102-fix-experimental

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

**Comment by tristan957** [comment]
I don't like this solution. It will break both Doxygen and hse-python.

**Comment by beckerg** [comment]
With this patch the code known as "experimental" is now a first class citizen within hse library (i.e., it is always built as part of hse lib, and the interfaces are always exported because you cannot selectively hide/expose interfaces via a shared library to select clients of the library).  This means that HSE_EXPERIMENTAL has gone away.  Instead there is a new file include/hse/hse_experimental.h which defines the "experimental" interfaces and makes clear that it is not part of the HSE library public API.  Clients who wish to use those interfaces can include hse_experimental.h at their own risk.
There remain two clients that need to be fixed, db_bench and hse-python, and so there's a workaround in hse.h to include hse_experimental.h for those clients only.  After they are updated we can remove the workaround.

**Comment by beckerg** [comment]
The old solution was a bunch of smoke an mirrors.  This solution is pretty clean, although I would prefer that all the "experimental" interfaces live in hse_experimenta.h (i.e., the types and flags, but that's a can of worms).
The fact is we always want build this stuff for our use, so using -DHSE_EXPERIMENTAL (which must always be defined to avoid compiler errors all over the place) to control building it is an illusion.  With this aproach we can also get rid of the -Dexperimental noise, which users get by default anyway..

**Comment by beckerg** [comment]
What's the doxygen issue?

**Comment by tristan957** [comment]
We explicitly made the decision to move the experimental interfaces inline to the other header files guarded by HSE_EXPERIMENTAL prior to 2.0.

docs/doxygen/meson.build:4

**Comment by beckerg** [comment]
It's just noise that no normal developer is going to want to see, and clearly we don't want them to use.  Are we now operating under some rule that says once we made a bad decision we cannot rectify it?  FWIW, I would have advised against that decision had I been aware of it, had to stay abreast of all the decisions being made as of late...

**Comment by tristan957** [comment]
I am not understanding how it was a bad decision.

**Comment by alexttx** [comment]
Pros and Cons of approach in this PR vs prior approach?

Pro:
- No ifdefs in code.
- Easier to maintain.

Pro or Con depending on viewpoint:
- Experimental APIs always visible, so user must take care to avoid them vs simply not being available depending on the build.

Con:
- ?

**Comment by tristan957** [comment]
A potential pro of this approach assuming -Dexperimental stays, which it definitely should, is that you don't install the experimental header if -Dexperimental=false.

**Comment by beckerg** [comment]
We cannot support them for internal use and yet hide them from the user.  It was an illusion that we could do so previous to my changes.  So nothing has changed.  The only thing we could do is hide hse_experimental.h somewhere deeper in the tree so that someone looking into include/hse doesn't see it.  I don't really care either way on that aspect...

File: cli/cli_util.c [hunk 1]
```diff
@@ -15,13 +15,8 @@
...
 #include <hse_util/parse_num.h>
 
-#ifndef HSE_EXPERIMENTAL
-#define HSE_EXPERIMENTAL
 #include <hse/hse.h>
-#undef HSE_EXPERIMENTAL
-#else
-#include <hse/hse.h>
-#endif
+#include <hse/hse_experimental.h>
```
Comment (tristan957) [inline cli/cli_util.c:19]: Rename it to hse/experimental.h

File: cli/meson.build [hunk 2]
```diff
@@ -15,11 +15,8 @@ hse_cli_sources = files(
     'storage_profile.c',
 )
 
-hse_cli_c_args = []
-
-if get_option('experimental')
-    hse_cli_c_args += '-DHSE_EXPERIMENTAL'
-endif
+hse_cli_c_args = [
```
Comment (tristan957) [inline cli/meson.build:18]: If this is an empty array, please remove it.

File: include/hse/hse_experimental.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#pragma GCC visibility push(default)
+
+
+/** @brief Request a data compaction operation.
```
Comment (tristan957) [inline include/hse/hse_experimental.h:31]: Look in types.h for doxygen addtogroup directives. You will want to use those here.

File: lib/binding/kvdb_interface.c [hunk 4]
```diff
@@ -775,6 +776,84 @@ hse_kvs_delete(
     return err;
 }
 
+uint64_t
```
Comment (tristan957) [inline lib/binding/kvdb_interface.c:779]: Where did this come from? Needs to return an hse_err_t.

File: tests/stress/meson.build [hunk 5]
```diff
@@ -173,7 +173,6 @@ foreach p, p_params: programs
         p,
         ['@0@.c'.format(p), 'stress_util.c'],
         c_args: [
-            '-DHSE_EXPERIMENTAL',
```
Comment (tristan957) [inline tests/stress/meson.build:176]: Lose the empty array

File: lib/meson.build [hunk 6]
```diff
@@ -81,13 +81,8 @@ hse_sources = [
 ]
 
 hse_c_args = [
```
Comment (tristan957) [inline lib/meson.build:83]: Lose the empty array

File: tests/unit/meson.build [hunk 7]
```diff
@@ -56,7 +56,6 @@ components = {
             },
         },
         'csched_test': {
```
Comment (tristan957) [inline tests/unit/meson.build:58]: Move the dictionary into one line

File: tools/meson.build [hunk 8]
```diff
@@ -191,7 +191,6 @@ tools = {
             'parm_groups.c',
         ),
         'c_args': [
```
Comment (tristan957) [inline tools/meson.build:193]: Lose the empty array

**Comment by beckerg** [comment]
With Tristan's suggestion (-Dexperimental=false does not install the header) I don't really see any Cons..

File: lib/binding/kvdb_interface.c (hunk 4 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:779]: Oh I see. This is a good change. I hated the experimental interface c file anyways.

**Review by tristan957 — APPROVED** [review]

**Comment by tristan957** [comment]
I have already started a PR for adding a stable build check to the CI

**Comment by tristan957** [comment]
As for what to do about experimental flags and types, we could do an experimental subfolder

hse/experimental/hse.h
hse/experimental/types.h
hse/experimental/flags.h

Alternatively just move them into the one experimental header.

**Comment by beckerg** [comment]
Cool, thanks!  Lemme make all those changes, wasn't sure if you wanted me to leave the empty meson args or not...  I'm inclined to move all the "experimental" interfaces (including types and flags) into experimental.h to keep them contained.  But that requires I add experimental.h to a few more files, no biggie...

**Comment by tristan957** [comment]
My approval is also contingent on you submitting the fix for hse-python :). It is a pretty small change. Look in hse2/hse.pxd. Grep for 'hse/hse.h'. Look at how there are declaration within the block. Copy the experimental ones out to a block under hse/experimental.h. Do that for all headers where experimental interfaces were previously inlined. You will an all caps `IF`. Guard your `hse/experimental.h` declarations within the IF.

**Comment by beckerg** [comment]
I can do that, but it leaves a window where hse-python wont build until these changes go into effect.  Same for db_bench.  The way I have it set up allows this patch to go in, then I can circle back and fix hse-python and db_bench after this hits master....

**Comment by tristan957** [comment]
Sounds good.

**Review by nabeelmmd — APPROVED** [review]

File: include/hse/experimental.h [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    uint64_t allocated_bytes; /**< allocated storage space for a kvdb */
+    uint64_t used_bytes;      /**< used storage space for a kvdb */
+};
+
+/**@} KVDB */
+
+/** @addtogroup KVS
+ * @{
+ */
+
```
Comment (tristan957) [inline include/hse/experimental.h:62]: Add prefix probe right under the enum definition

File: include/hse/experimental.h [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+ * @param kvdb: KVDB handle from hse_kvdb_open().
+ * @param flags: Compaction flags.
+ *
+ * @remark @p kvdb must not be NULL.
+ * @remark @p flags must not be 0.
+ *
+ * @returns Error status.
+ */
+hse_err_t
+hse_kvdb_compact(struct hse_kvdb *kvdb, unsigned int flags);
```
Comment (tristan957) [inline include/hse/experimental.h:100]: I would say just have the one KVDB doxygen section (ie merge the flags, functions, and types)

**Review by tristan957 — CHANGES_REQUESTED** [review]

**Review by tristan957 — APPROVED** [review]

**Pull request merged**